### ISSUE TITLE
ANSI colors in OutputChannel

### ISSUE THREAD

**Issue by seanmcbreen** [microsoft/vscode#571]
_From @SamVerschueren on October 30, 2015 6:9_

I might have seen a post regarding styling of the `OutputChannel`, but not sure though.

I think it would be a nice feature to support ANSI colors. This way libraries like [chalk](https://github.com/chalk) can be used to style the output. Another benefit is that if you develop a plugin like Yeoman, who outputs to the console, the output can be directly passed to the `OutputChannel`.

Not sure if this is possible, but I think it would be a nice feature. No need to implement our own styling stuff in the channel, just use libraries that already exist.

_Copied from original issue: Microsoft/vscode-extensionbuilders#69_

**Comment by seanmcbreen** [comment]
_From @bpasero on October 30, 2015 7:16_

Yes, this is on the plan for post November.

**Comment by seanmcbreen** [comment]
_From @SamVerschueren on October 30, 2015 7:16_

Awesome, great job!

**Comment by daviwil** [comment]
Excellent!

**Comment by orta** [comment]
As there's not been too much activity on this thread, I'd like to drop a +1 - I was comparing some of [my extension's](https://github.com/orta/vscode-jest/) output from the debug console to it's user's Output Channel and would love to be able to show colour in there.

![screen shot 2016-11-29 at 10 46 42](https://cloud.githubusercontent.com/assets/49038/20706857/357811ea-b621-11e6-93d3-8ef101d8aebf.png)

![screen shot 2016-11-29 at 10 46 45](https://cloud.githubusercontent.com/assets/49038/20706873/4ea0b4b0-b621-11e6-843a-d3aa53260a6a.png)

**Comment by bartlewis** [comment]
+1

**Comment by abotsi** [comment]
Up! 
Is there anything new on this side? Do you have an idea of why a such feature cannot be added?

**Comment by orta** [comment]
To provide a workaround, you can [create a custom log syntax](https://twitter.com/wallabyjs/status/827301961867288576) for the output  - see https://github.com/IBM-Bluemix/vscode-log-output-colorizer

**Comment by abotsi** [comment]
Thank you @orta for your quick response, but the extension you proposed did't enhances my situation. Actually, my problem is I don't get anything from [ProgressBarPlugin](https://github.com/clessg/progress-bar-webpack-plugin) when I bundle my project with WebPack with VSCode task, whereas I get a nice print when I run the command from a terminal...

_Output panel (actual)_ :
![output panel - actual](https://cloud.githubusercontent.com/assets/9976899/22835135/02cb746e-efb8-11e6-8971-00d53267b767.png)

_Terminal panel (expected)_ :
![terminal panel - expected](https://cloud.githubusercontent.com/assets/9976899/22835160/1c957962-efb8-11e6-98a7-b16850377652.png)

I don't know if it's directly related with issue (this plugin WebPack use [chalk](https://github.com/chalk/chalk), which could be the problem), maybe I should create another thread...

**Comment by orta** [comment]
I was noting that you could write a colorizer yourself, not that it would help you right now.

**Comment by erichiller** [comment]
+1 this - seems to be possible to render the ansi in the same method as the internal termimal - in fact I would prefer it this way! Because then **my variables and other environment configs I define within that environment become available**.

**Comment by seanpoulter** [comment]
Is anyone on the team able to review a PR for this if I put one together? If yes, any tips on which files you're expecting to change?

**Comment by thw0rted** [comment]
Hello from the future! What's the status of this issue?  I just started using Maven after a few years out of the Java ecosystem, and apparently now it outputs color codes.  It works in `cmd` and PowerShell, but I get literal color escape codes in Git Bash and the VSCode "Output" pane.  I'm just trying to work out if this is a VSCode problem, a Git Bash problem, or maybe both.

**Comment by yourhead** [comment]
Hi,

I've read through all the details here and potential workarounds to support specific languages -- but is there any way to just support ANSI colors?

Currently if I build an extension that runs a command that produces ANSI colored text there does not seem to be a way to show it to the user.  My options seem to be:

- Send to a terminal (which does support ANSI colors).  But text is sent to stdin -- so it will try to run all of my output. 👎 
- Send text to an output channel. This works nicely but there is no support for ANSI colors

The above mentioned workarounds for coloring the output based on regex language definitions is nice -- but wouldn't ANSI color characters still show up as garbage characters?


In summary:  I need a read-only view for sending ANSI colored text to. Is there a way to do that? If not please add that -- I think it would be a nice addition. Thx.

**Comment by Gruntfuggly** [comment]
It would be nice to be able to insert colour. At the moment I'm resorting to adding text decorations, which works but feels like overkill...

**Comment by gustavooferreira** [comment]
Hey guys, any updates on this?

**Comment by Phrogz** [comment]
I have a Lua script that outputs ANSI; when I use Code Runner extension to run my script, I would love to see color in the Output panel instead of escape sequences.

**Comment by oh-hello-tomcat** [comment]
Ginkgo's output is so neat and tidy. The missing ANSI code support takes away a lot of that. Just curious: is it really that hard to do?

**Comment by jrieken** [comment]
No changes are planned for the output channels but there is terminal and pseudo terminal API

**Comment by felixbuenemann** [comment]
I vote for re-opening this issue.

@jrieken Why did you label this issue as out-of-scope?

**Comment by vpillinger** [comment]
Not have colorized output breaks a lot of packages, but my main concern is that any sort of integrated unit testing or test-driven development runner is broken if the test output includes colorized results.

By default and sensibility most test frameworks output colorized results, so this is a big deal breaker for me personally.

**Comment by jbcpollak** [comment]
pytest output is next to impossible to debug without this feature. Please consider re-opening it. For now I've resorted to running pytest on the terminal.

**Comment by sn0wf4ke** [comment]
yes. for pytest it would be a great enhancement on usability.

**Comment by paxcodes** [comment]
@jbcpollak @sn0wf4ke I also use Pytest and would love to have this feature. There is an existing issue over at vscode-python repo. https://github.com/microsoft/vscode-python/issues/12615

Giving the issue a thumbs up would help as one of the maintainers say,

> how many 👍 votes the opening comment gets [..] help us make our decision.

**Comment by bconway** [comment]
I would love to take this out of my code:
```
// VSCode does not support colors in the output channel:
// https://github.com/Microsoft/vscode/issues/571
NoColor:    true,
```

**Comment by nrhoffmann** [comment]
I'd recommend using an extension pseudoterminal.

 Documentation is a little scarce. See [ShMcK/vscode-pseudoterminal](https://github.com/ShMcK/vscode-pseudoterminal) for a simple example of this in use.

**Comment by malmod** [comment]
Please reopen this!

**Comment by OliverScherf** [comment]
We really need this! Reopen please.

**Comment by Athayde3586** [comment]
No

**Comment by itsmeboy0** [comment]
+1 for this!

**Comment by Bi0max** [comment]
+1 for this!!

**Comment by Rickaym** [comment]
+1 for this!

**Comment by m-paternostro** [comment]
+1 for this!

**Comment by Tantol** [comment]
+1

**Comment by JamyDev** [comment]
@isidorn Would you be able to bring this issue up for reconsideration? I know it's an old one, but we're integrating with Test Explorer and that uses the OutputChannel for our build+test output. Would be nice to have ANSI support.

**Comment by isidorn** [comment]
@JamyDev would be great if you can open a new issue. I am not sure if this is something we will tackle any time soon though.

**Comment by MohamedLamineAllal** [comment]
I made a year ago and half a proof of concept. That tackles cleaning the log. A day work or something. And since then it was buried in my laptop. Two days ago i went through the repo. And got curious. I thought the problem would have been fixed with a newer vscode version. Surprised it's not yet ❗

So i decided. I'll try to publish something workable, may be some will find it helpful. And not leave it buried on my laptop. From willing to just publish something. I ended by working out a fully optimized extension. That works and tackles the problem well.

Check it out
- [TheMagicianDev.vsc-clean-output-log ( CleanLog )](https://marketplace.visualstudio.com/items?itemName=TheMagicianDev.vsc-clean-output-log)
- https://github.com/TheMagicianDev/vsc-clean-output-log

**Comment by MohamedLamineAllal** [comment]
I did some extra validation. And i see myself able to fully support ansi colors codes colorization. I'll try to handle that on the weekend.
Idea
- Use editor decoration. All Output panels are editors. And are accessible through `vscode.window.visibleTextEditors`, Fully validated
<img width="549" alt="image" src="https://github.com/microsoft/vscode/assets/20062543/3272c184-e212-4557-93ad-74c2aa1f4fdf">

- I just need to check and validate Output channel change event (editor something) and the `ranges` handling.

**Comment by iBicha** [comment]
Please reopen this issue - the need for this will not magically disappear, thanks

**Comment by ffMathy** [comment]
@jrieken could you provide some justification as to why this was closed as out-of-scope? There are many who still need it, so an explanation would be great.
